app/src/main/java/com/andela/art/Login/LoginActivity.java

```diff
@@ -0,0 +1,175 @@
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        mAuth.addAuthStateListener(mAuthListener);
+    }
+
+    @Override
+    public void onStop() {
+        super.onStop();
+        if (mAuthListener != null) {
+            mAuth.removeAuthStateListener(mAuthListener);
+        }
+    }
+
+    @Override
+    protected void attachBaseContext(Context newBase) {
+        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        CalligraphyConfig.initDefault(new CalligraphyConfig.Builder()
+                        .setDefaultFontPath("fonts/Roboto-Regular.ttf")
+                        .setFontAttrId(R.attr.fontPath)
+                        .build()
+        );
+        setContentView(R.layout.google_login);
+        Window w = getWindow();
+        w.setFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS, WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
+        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
+        button = findViewById(R.id.googleSignInButton);
+
+        button.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                signIn();
+            }
+        });
+
+        // Instance of mAuth
+        mAuth = FirebaseAuth.getInstance();
+
+        mAuthListener = new FirebaseAuth.AuthStateListener() {
+            @Override
+            public void onAuthStateChanged(@NonNull FirebaseAuth firebaseAuth) {
+                 if (firebaseAuth.getCurrentUser() != null){
+                     // Hide the progress dialog if its showing.
+                     mConnectionProgressDialog.dismiss();
+                     Intent redirect = new Intent(LoginActivity.this, NavigationDrawerActivity.class);
+                     LoginActivity.this.startActivity(redirect);
+                     // User is signed in
+                     Log.d(TAG, "onAuthStateChanged:signed_in:" + firebaseAuth.getCurrentUser().getUid());
+                 }
+            }
+        };
+
+        // Configure Google Sign In
+        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+                .requestIdToken(getString(R.string.default_web_client_id))
+                .requestEmail()
+                .requestProfile()
+                .build();
+
+        // Build a GoogleSignInClient with the options specified by gso.
+        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);
+
+        // Configure the ProgressDialog that will be shown if there is a
+        // delay in presenting the user with the next sign in step.
+        mConnectionProgressDialog = new ProgressDialog(this);
+        mConnectionProgressDialog.setMessage("Signing in...");
+    }
+
+  @Override public void firebasewithGoogleAuth(GoogleSignInAccount account) {
```

Comment:
This could be done in another layer and then injected into the activity so that the view remains dumb.

Reply:
Hey @zmwas in this instance this method has a listener on it. Is there a way of having listeners in a presenter?

Reply:
You could create a class to do the login and then inject the class into the presenter. When login is called on the activity the presenter just uses the injected class and calls the sign in. That way the firebase code could be mocked as it will be injected in the constructor and the activity will remain dumb.

Reply:
Noted, that might work, will try this tomorrow.